File: Solutions/22-generate-parentheses/generate-parentheses.go
var combinations []string
var number int

func generateParenthesis(n int) []string {
	combinations = []string{}
	number = n

	backtrack("", 0, 0)

	return combinations
}

func backtrack(current string, open, close int) {
	if len(current) == number*2 {
		combinations = append(combinations, current)
		return
	}

	if open < number {
		backtrack(current+"(", open+1, close)
	}

	if close < open {
		backtrack(current+")", open, close+1)
	}
}

File: Solutions/22-generate-parentheses/generate-parentheses.ts
function generateParenthesis(n: number): string[] {
    const ans: string[] = [];
    
    function solve(s: string, open: number, close: number): void {
        if (close === n && open === n) {
            ans.push(s);
            return;
        }
        
        if (open < n) {
            solve(s + "(", open + 1, close);
        }
        
        if (close < open) {
            solve(s + ")", open, close + 1);
        }
    }
    
    solve("", 0, 0);
    return ans;
}